lib/application/widgets/theme/tags_field.dart

```diff
@@ -0,0 +1,273 @@
+    void onTagTap(String tag) {
+      controller.tags = controller.tags..remove(tag);
+      tags.value = controller.tags;
+    }
+
+    final tagsWidgets = controller.tags.map((tag) => _SelectedTag(tag: tag, onTap: onTagTap));
+
+    final textField = TextField(
+      focusNode: focus,
+      controller: fieldController,
+      cursorColor: theme.secondarySwatch,
+      autocorrect: false,
+      enableSuggestions: false,
+      decoration: InputDecoration(
+        isCollapsed: true,
+        contentPadding: EdgeInsets.zero,
+        fillColor: Colors.transparent,
+        hintText: strings.tagsHint,
+        hintStyle: textTheme.bodyText2?.copyWith(color: theme.neutralSwatch.shade600),
+      ),
+      textAlignVertical: TextAlignVertical.center,
+      style: textTheme.bodyText2,
+      keyboardType: TextInputType.text,
+      textInputAction: TextInputAction.done,
+      inputFormatters: [_TagFieldFormatter()],
+      onSubmitted: (term) {
+        fieldController?.clear();
+        onSubmitted?.call(term);
+        focus?.requestFocus();
+      },
+    );
+
+    return Wrap(
+      spacing: context.rawSpacing(Spacing.small),
+      runSpacing: context.rawSpacing(Spacing.small),
+      children: [
+        ...tagsWidgets,
+        IntrinsicWidth(child: textField),
+      ],
+    );
+  }
+}
+
+class _SelectedTag extends ConsumerWidget {
+  const _SelectedTag({required this.tag, this.onTap});
+
+  final String tag;
+
+  /// Called when the user taps in the respective tag.
+  final void Function(String tag)? onTap;
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final theme = ref.watch(themeController);
+    final textTheme = Theme.of(context).textTheme;
+
+    return Material(
+      child: InkWell(
+        onTap: () => onTap?.call(tag),
+        child: Ink(
+          decoration: BoxDecoration(
+            borderRadius: dimens.textTagBorderRadius,
+            color: theme.secondarySwatch.shade600,
+          ),
+          child: Row(
+            mainAxisSize: MainAxisSize.min,
+            children: [
+              Text(tag, style: textTheme.caption),
+              context.horizontalBox(Spacing.xxxSmall),
+              Image.asset(images.closeAsset, height: dimens.tagsRemoveIconSize, width: dimens.tagsRemoveIconSize),
+            ],
+          ).withAllPadding(context, Spacing.xxSmall),
+        ),
+      ),
+    );
+  }
+}
+
+/// Formats the text from [_TagsTextField] `TextField`.
+class _TagFieldFormatter extends TextInputFormatter {
+  static const _maxTagLength = 15;
+
+  @override
+  TextEditingValue formatEditUpdate(TextEditingValue oldValue, TextEditingValue newValue) {
+    if (newValue.text.length > _maxTagLength) {
+      return TextEditingValue(text: oldValue.text, selection: oldValue.selection);
+    }
+
+    // Restrict the content to only accept alphanumeric space, underscore and comma characters.
+    final alphanumRegex = RegExp(r'^[a-zA-Z0-9_ ,]*$');
```

Comment:
ditto: shouldn't this be a value provided by some validation class? This looks to me like a leaked business definition